### PR TITLE
Add MINIFS and MAXIFS functions

### LINKED ISSUE #1049: MINIFS nad MAXIFS functions
### Description

Implement MINIFS nad MAXIFS functions

### Links

- https://github.com/handsontable/hyperformula/pull/1047

### PR DESCRIPTION
**Pull request by sequba**
### Context

@rkretzschmar in #1047 wrote:
> - adds two not implemented functions `MINIFS` and `MAXIFS` to the built-in functions
> - adds only English and German translations, all other language files contain the English translation to pass the `i18n.spec.ts`
> - describes the two new functions in the `built-in-functions.md`
> - added two new test suites `function-minifs.spec.ts` and `function-maxifs.spec.ts`

TODO:
- [x] check compatibility
  - [x] ODFF (no such function)
  - [x] Excel
  - [x] Google Sheets
- [x] refactor
  - [x] rename `SumifPlugin` -> `ConditionalAggregationPlugin`
  - [x] refactor existing functions inside  `ConditionalAggregationPlugin`
  - [x] define MAXIFS and MINIFS inside `ConditionalAggregationPlugin`
- [x] add missing translations
- [x] don't throw error when taking min/max of zero values
- [x] test with other data types (currency, date etc)

Co-authored-by: Rene Kretzschmar <rkretzschmar@gmail.com>

### How has this been tested?

Unit tests pass (including new test suites for MINIFS and MAXIFS

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Additional language file or change to the existing one (translations)
- [ ] Change to the documentation

### Related issues:
Fixes #1049 
#1047 

### Checklist:

- [x] My code follows the code style of this project.
- [x] I described the modification in the CHANGELOG.md file.
- [x] My change requires a change to the documentation.
- [ ] My change requires a migration guide.

### CODE REVIEW

File: src/interpreter/plugin/ConditionalAggregationPlugin.ts [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+  }
+
+  if (isExtendedNumber(arg)) {
+    return getRawValue(arg)
+  }
+
+  return undefined
+}
+
+export class ConditionalAggregationPlugin extends FunctionPlugin implements FunctionPluginTypecheck<ConditionalAggregationPlugin> {
```
Comment (sequba) [inline src/interpreter/plugin/ConditionalAggregationPlugin.ts:77]: I renamed the class `SumifPlugin` to `ConditionalAggregationPlugin`. It may be used as an argument for `registerFunctionPlugin()` and `unregisterFunctionPlugin()` but its never mentioned in our docs so I consider this as non-breaking change. @warpech WDYT?

File: src/interpreter/plugin/NumericAggregationPlugin.ts [hunk 2]
```diff
@@ -628,7 +628,7 @@ export class NumericAggregationPlugin extends FunctionPlugin implements Function
 
     const rangeStart = range.start
     const rangeEnd = range.end
-    const rangeVertex = this.dependencyGraph.getRange(rangeStart, rangeEnd)!
+    const rangeVertex = this.dependencyGraph.getRange(rangeStart, rangeEnd)
```
Comment (sequba) [inline src/interpreter/plugin/NumericAggregationPlugin.ts:631]: Fixing linter warning

File: src/interpreter/plugin/NumericAggregationPlugin.ts [hunk 3]
```diff
@@ -666,7 +666,12 @@ export class NumericAggregationPlugin extends FunctionPlugin implements Function
   private getRangeValues<T>(functionName: string, range: AbsoluteCellRange, mapFunction: MapOperation<T>, coercionFunction: coercionOperation): (T | CellError)[] {
     const rangeResult: (T | CellError)[] = []
     const {smallerRangeVertex, restRange} = this.dependencyGraph.rangeMapping.findSmallerRange(range)
-    const currentRangeVertex = this.dependencyGraph.getRange(range.start, range.end)!
+    const currentRangeVertex = this.dependencyGraph.getRange(range.start, range.end)
+
+    if (currentRangeVertex === undefined) {
+      throw new Error(`Range does not exists in graph: (${range.start}, ${range.end})`)
+    }
+
```
Comment (sequba) [inline src/interpreter/plugin/NumericAggregationPlugin.ts:674]: Fixing linter warning

File: src/interpreter/plugin/ConditionalAggregationPlugin.ts (hunk 1 above)
Comment (warpech) [reply src/interpreter/plugin/ConditionalAggregationPlugin.ts:77]: Yeah, I think it is an internal change.

File: docs/guide/built-in-functions.md [hunk 4]
```diff
@@ -308,7 +302,7 @@ Total number of functions: **{{ $page.functionsCount }}**
 | SUBTOTAL        | Computes aggregation using function specified by number.                                                      | SUBTOTAL(Function; Number1; Number2; ... Number30)                                          |
 | SUM             | Adds all the numbers in a range of cells.                                                                     | SUM(Number1; Number2; ...; Number30)                                                        |
 | SUMIF           | Adds the cells specified by given criteria.                                                                   | SUMIF(Range; Criteria; Sumrange)                                                            |
-| SUMIFS          | Returns the sum of the values of cells in a range that meets multiple criteria in multiple ranges.            | SUMIFS(Sum_Range ; Criterion_range1 ; Criterion1 [ ; Criterion_range2 ; Criterion2 [;...]]) |
+| SUMIFS          | Returns the sum of the cells in a range that meet a set of criteria.                                          | SUMIFS(Sum_Range ; Criterion_range1 ; Criterion1 [ ; Criterion_range2 ; Criterion2 [;...]]) |
```
[the remark this answers is not in the document]
Comment (sequba) [reply docs/guide/built-in-functions.md:305]: The sentence might not be clear enough, but the second clause "that meet a set of criteria" corresponds to the noun cells (plural -> "meet").

I tried to think of a way to rephrase this sentence so it so more clear, but I couldn't find a better option.

File: docs/guide/built-in-functions.md [hunk 5]
```diff
@@ -349,165 +343,167 @@ Total number of functions: **{{ $page.functionsCount }}**
...
+| LOGNORMINV      | Returns value of inverse lognormal distribution.                                                          | LOGNORMINV(P; Mean; Stddev)                                                                 |
+| LOGINV          | Returns value of inverse lognormal distribution.                                                          | LOGINV(P; Mean; Stddev)                                                                     |
+| MAX             | Returns the maximum value in a list of arguments.                                                         | MAX(Number1; Number2; ...Number30)                                                          |
+| MAXA            | Returns the maximum value in a list of arguments.                                                         | MAXA(Value1; Value2; ... Value30)                                                           |
+| MAXIFS          | Returns the maximum of the cells in a range that meet a set of criteria.                                  | MAXIFS(Max_Range ; Criterion_range1 ; Criterion1 [ ; Criterion_range2 ; Criterion2 [;...]]) |
```
[the remark this answers is not in the document]
Comment (sequba) [reply docs/guide/built-in-functions.md]: "cells" (plural) "meet a set of criteria"

I tried to think of a way to rephrase this sentence so it so more clear, but I couldn't find a better option.

File: docs/guide/built-in-functions.md [hunk 6]
```diff
@@ -349,165 +343,167 @@ Total number of functions: **{{ $page.functionsCount }}**
...
+| MEDIAN          | Returns the median of a set of numbers.                                                                   | MEDIAN(Number1; Number2; ...Number30)                                                       |
+| MIN             | Returns the minimum value in a list of arguments.                                                         | MIN(Number1; Number2; ...Number30)                                                          |
+| MINA            | Returns the minimum value in a list of arguments.                                                         | MINA(Value1; Value2; ... Value30)                                                           |
+| MINIFS          | Returns the minimum of the cells in a range that meet a set of criteria.                                  | MINIFS(Min_Range ; Criterion_range1 ; Criterion1 [ ; Criterion_range2 ; Criterion2 [;...]]) |
```
[the remark this answers is not in the document]
Comment (sequba) [reply docs/guide/built-in-functions.md]: "cells" (plural) "meet a set of criteria"

I tried to think of a way to rephrase this sentence so it so more clear, but I couldn't find a better option.

[the remark this answers is not in the document]
Comment (rkretzschmar) [reply docs/guide/built-in-functions.md]: What about:

> The MINIFS function returns the minimum value among cells specified by a given set of conditions or criteria.

like at https://support.microsoft.com/en-us/office/minifs-function-6ca1ddaa-079b-4e74-80cc-72eef32e6599

File: src/i18n/languages/deDE.ts [hunk 7]
```diff
@@ -133,12 +133,14 @@ const dictionary: RawTranslationPackage = {
     MATCH: 'VERGLEICH',
     MAX: 'MAX',
     MAXA: 'MAXA',
+    MAXIFS: 'MAXWENNS',
```
[the remark this answers is not in the document]
Comment (sequba) [reply src/i18n/languages/deDE.ts:136]: I was using Excel Function Translator. I described it here: https://kb.handsontable.com/display/PH/Introduction (internal link)

File: CHANGELOG.md [hunk 8]
```diff
@@ -6,6 +6,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+- Added MAXIFS and MINIFS functions. Co-authored-by: Rene Kretzschmar <rkretzschmar@gmail.com>. [#1049](https://github.com/handsontable/hyperformula/issues/1049)
```
[the remark this answers is not in the document]
Comment (sequba) [reply CHANGELOG.md]: You're right. I removed this information from the changelog.

**Comment by sequba** [comment]
@rkretzschmar For your information:

Your general idea for the implementation of these two functions was good. I only made a few changes to comply with the standards of HyperFormula:
- move the implementation of MAXIFS and MINIFS to different file and refactored it
- return `0` instead of an error when trying to calculate max/min of an empty range (compatibility with Microsoft Excel)
- make it ignore empty cells, strings and booleans (compatibility with Microsoft Excel)
- add unit tests for handling other data types and some edge cases
- add missing translations
- rephrase the functions' descriptions in the documentation
- add changelog entry
- pull recent changes from `develop` branch

Once again, thanks for your contribution. You helped us a lot :)

File: docs/guide/built-in-functions.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (sequba) [reply docs/guide/built-in-functions.md]: @rkretzschmar It's a good point. The problem is, we cannot copy content from Microsoft documentation due to copyright issues.

File: docs/guide/built-in-functions.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (kirszenbaum) [reply docs/guide/built-in-functions.md:305]: @sequba @warpech – if the SUM, SUMIF, and SUMIFS functions work the same as in Excel and Sheets (I suppose they do, as they're not listed [here](https://hyperformula.handsontable.com/guide/known-limitations.html#google-sheets-and-microsoft-excel)), I'd describe them like this:

- SUM: Sums up the values of the specified cells.
- SUMIF: Sums up the values of all cells that belong to the specified range and meet the specified condition.
- SUMIFS: Sums up the values of all cells that belong to the specified range and meet the specified sets of conditions.

The descriptions for SUMIF and SUMIFS are simplified, but to explain all possible use cases, we'd need to separately describe each argument (Range; Criteria; Sumrange) separately, provide some examples etc.

[the remark this answers is not in the document]
Comment (sequba) [reply docs/guide/built-in-functions.md]: @kirszenbaum These look perfect. I updated the descriptions according to your suggestions. Thanks a lot for your help.

File: docs/guide/built-in-functions.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (warpech) [reply docs/guide/built-in-functions.md]: I get it now, thanks for explaining. I think it is good actually

File: docs/guide/built-in-functions.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (warpech) [reply docs/guide/built-in-functions.md]: Let's keep the current text then

File: src/i18n/languages/deDE.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (warpech) [reply src/i18n/languages/deDE.ts:136]: Thanks!

**Review by warpech — APPROVED** [review]
I reviewed the code and it looks good. Thanks for the code reuse in `ConditionalAggregationPlugin.ts`!
